### PR TITLE
Add maxIGSize to NodePool constructor, instead of using global flag

### PR DESCRIPTION
**Pull request by panslava**
This is refactoring of NodePool, following from this https://github.com/kubernetes/ingress-gce/pull/1707

### CODE REVIEW

File: pkg/flags/flags.go [hunk 1]
```diff
@@ -244,8 +244,8 @@ L7 load balancing. CSV values accepted. Example: -node-port-ranges=80,8080,400-5
 	flag.StringVar(&F.GKEClusterType, "gke-cluster-type", "ZONAL", "The cluster type of the GKE cluster this Ingress Controller will be interacting with")
 	flag.BoolVar(&F.EnableTrafficScaling, "enable-traffic-scaling", false, "Enable support for Service {max-rate-per-endpoint, capacity-scaler}")
 	flag.BoolVar(&F.EnableEndpointSlices, "enable-endpoint-slices", false, "Enable using Endpoint Slices API instead of Endpoints API")
-	flag.BoolVar(&F.EnableMultipleIgs, "enable-multiple-igs", false, "Enable using unmanaged instance group management")
-	flag.IntVar(&F.MaxIgSize, "max-ig-size", 1000, "Max number of instances in Instance Group")
+	flag.BoolVar(&F.EnableMultipleIGs, "enable-multiple-igs", false, "Enable using multiple unmanaged instance groups")
+	flag.IntVar(&F.MaxIGSize, "max-ig-size", 1000, "Max number of instances in Instance Group")
```
Comment (bowei) [inline pkg/flags/flags.go:248]: Not clear to me why we need to expose this as a flag. We can still keep it as var in the pool for testing purposes, but are we really going to change this constant?

File: pkg/instances/instances.go [hunk 2]
```diff
@@ -47,22 +46,33 @@ type Instances struct {
...
+	Cloud      InstanceGroups
+	Namer      namer.BackendNamer
+	Recorders  recorderSource
+	BasePath   string
+	ZoneLister ZoneLister
+	MaxIGSize  int
+}
+
+// NewNodePool creates a new node pool using NodePoolConfig.
+func NewNodePool(config NodePoolConfig) NodePool {
```
Comment (bowei) [inline pkg/instances/instances.go:68]: take a pointer here unless you really want to copy things

File: pkg/instances/instances.go [hunk 3]
```diff
@@ -337,19 +347,19 @@ func (i *Instances) Sync(nodes []string) (err error) {
...
 		// Individual InstanceGroup has a limit for 1000 instances in it.
 		// As a result, it's not possible to add more to it.
-		if len(kubeNodes) > flags.F.MaxIgSize {
+		if len(kubeNodes) > i.maxIGSize {
 			// List() will return a sorted list so the kubeNodesList truncation will have a stable set of nodes.
 			kubeNodesList := kubeNodes.List()
 
 			// Store first 10 truncated nodes for logging
-			truncatedNodesSample := kubeNodesList[flags.F.MaxIgSize:]
+			truncatedNodesSample := kubeNodesList[i.maxIGSize:]
```
Comment (bowei) [inline pkg/instances/instances.go:355]: suggestion:

truncateForLog := func(l []string) []string {
  if len(l) <= 10 { return l }
  return l[:10]
}

File: pkg/instances/instances.go (hunk 2 above)
Comment (panslava) [reply pkg/instances/instances.go]: Fixed

File: pkg/instances/instances.go (hunk 3 above)
Comment (panslava) [reply pkg/instances/instances.go]: Done

File: pkg/flags/flags.go (hunk 1 above)
Comment (panslava) [reply pkg/flags/flags.go:248]: We will use this for easier end2end testing, so on some testing environment we would be able to set this flag to 5 (or other small value), and then test that our multiple instance group management actually working with google cloud apis, but without creating 1000+ nodes

File: pkg/instances/instances_test.go [hunk 4]
```diff
@@ -84,9 +89,9 @@ func TestNodePoolSync(t *testing.T) {
 			defaultZone: {
 				ig: testCase.gceNodes,
 			},
-		})
+		}, maxIGSize)
```
Comment (cezarygerard) [inline pkg/instances/instances_test.go:92]: could you create the map[string]IGsToInstances outside of the function call?
just make it a variable 
```
readableVarName := map[string]IGsToInstances{
			defaultZone: {
				ig: testCase.gceNodes,
			},
		}
fakeGCEInstanceGroups := NewFakeInstanceGroups(readableVarName, maxIGSize)
```

File: pkg/instances/instances_test.go [hunk 5]
```diff
@@ -144,12 +149,13 @@ func TestNodePoolSync(t *testing.T) {
 }
 
 func TestSetNamedPorts(t *testing.T) {
+	maxIGSize := 1000
 	fakeIGs := NewFakeInstanceGroups(map[string]IGsToInstances{
```
Comment (cezarygerard) [inline pkg/instances/instances_test.go:153]: same here,create asisgn the map to variable

File: pkg/instances/instances_test.go [hunk 6]
```diff
@@ -198,11 +204,12 @@ func TestSetNamedPorts(t *testing.T) {
 }
 
 func TestGetInstanceReferences(t *testing.T) {
+	maxIGSize := 1000
 	pool := newNodePool(NewFakeInstanceGroups(map[string]IGsToInstances{
```
Comment (cezarygerard) [inline pkg/instances/instances_test.go:208]: same here,
add some variables for readability

File: pkg/instances/instances_test.go (hunk 4 above)
Comment (panslava) [reply pkg/instances/instances_test.go:92]: Done

File: pkg/instances/instances_test.go (hunk 5 above)
Comment (panslava) [reply pkg/instances/instances_test.go]: Done

File: pkg/instances/instances_test.go (hunk 6 above)
Comment (panslava) [reply pkg/instances/instances_test.go]: Done

**Review by cezarygerard — APPROVED** [review]

**Pull request merged**